### ISSUE TITLE
[BUG]: 4.0.1 not compatible with @sentry/react-native 4.1.2

### ISSUE THREAD

**Issue by liamjones** [wix/sentry-testkit#113]
**Describe the bug**
Trying to upgrade to Sentry RN 4+ following the merging of PR #109. Typing now seems to be broken for the transport property of `Sentry.init()` and it fails to work.

**To Reproduce**

```typescript
import sentryTestkit from "sentry-testkit";
import * as Sentry from "@sentry/react-native";

const { testkit, sentryTransport } = sentryTestkit();

Sentry.init({
  dsn: "https://dummy_dsn_with_valid_format@example.org/123",
  enableNative: false,
  tracesSampleRate: 1,
  transport: sentryTransport // This line shows a type error
});
````

Type error:

> TS2322: Type 'new (options: any) => Transport' is not assignable to type '(transportOptions: BrowserTransportOptions) => Transport'.   Type 'new (options: any) => Transport' provides no match for the signature '(transportOptions: BrowserTransportOptions): Transport'.  options.d.ts(212, 5): The expected type comes from property 'transport' which is declared here on type 'ReactNativeOptions'

Upon running:

```
Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'bind')

      19 | const { testkit, sentryTransport } = sentryTestkit()
      20 |
    > 21 | Sentry.init({
         |        ^
      22 |     dsn: 'https://dummy_dsn_with_valid_format@example.org/123',
      23 |     enableNative: false,
      24 |     tracesSampleRate: 1,

      at Object.getNativeFetchImplementation (../../../node_modules/src/transports/utils.ts:77:1)
      at makeFetchTransport (../../../node_modules/src/transports/fetch.ts:12:1)
      at Object.transport (../../../node_modules/@sentry/react-native/src/js/client.ts:32:14)
      at new BaseClient (../../../node_modules/@sentry/src/baseclient.ts:108:1)
      at ReactNativeClient._createSuperInternal (../../../node_modules/@sentry/react-native/dist/js/client.js:34:311)
      at new ReactNativeClient (../../../node_modules/@sentry/react-native/src/js/client.ts:36:6)
      at initAndBind (../../../node_modules/@sentry/src/sdk.ts:34:1)
      at Object.init (../../../node_modules/@sentry/react-native/src/js/sdk.tsx:115:3)
```

**Expected behavior**
No typing error on the `transport` line, bootstrapping of `Sentry.init()` works.

**Sentry Versions (please complete the following information):**
 - Sentry: 4.1.2
 - Sentry-Testkit 4.0.1

**Comment by liamjones** [comment]
I _think_ this might be a Sentry issue, investigating further...

**Comment by liamjones** [comment]
Okay, this does look to be sentry-testkit specific rather than a @sentry/react-native issue.

A test with @sentry/react-native@3.4.3 doing a plain init in like this:

```typescript
  Sentry.init({
    dsn: 'https://dummy_dsn_with_valid_format@example.org/123',
    enableNative: false,
    tracesSampleRate: 1,
  });
```

will fail with `TypeError: Cannot read properties of undefined (reading 'bind')`.

Adding sentry-testkit@3.3.8's custom transport allows this to work:

```typescript
  const {sentryTransport} = sentryTestkit();

  Sentry.init({
    dsn: 'https://dummy_dsn_with_valid_format@example.org/123',
    enableNative: false,
    tracesSampleRate: 1,
    transport: sentryTransport,
  });
```

Moving to @sentry/react-native@4.1.2 without the transport results in the same error as @sentry/react-native@3.4.3 so their behaviour doesn't seem to have changed.

However, adding sentry-testkit@4.0.1's transport doesn't fix the error this time.

Any ideas @mlmmn? I've uploaded a minimal repro here: https://github.com/liamjones/SentryTestkit4ReactNativeBug (just `yarn` then `yarn test`)

**Comment by zivl** [comment]
@liamjones i'm confused, there's a TS typing error or broken functionality?

**Comment by liamjones** [comment]
@zivl sorry for the confusion; I originally thought it was just a typing issue and reported that but as I went further with the upgrade I discovered that the functionality is actually broken with @sentry/react-native

**Comment by zivl** [comment]
thanks @liamjones - I'll take a look at your repo and we'll see how to deal with it.
i'll update here once i have more insights...
@liamjones, @mlmmn - you are more than welcome to do the same and share here yours of course

**Comment by zivl** [comment]
@liamjones I think I understand the issue:

when `@sentry/react-native` is initialized it checks for the **native** transport layer (i.e. device network connection) and since in test environment it does not exist - it falls back to `@sentry/browser`'s `makeFetchTransport` function **directly** [(see here)](https://github.com/getsentry/sentry-react-native/blob/main/src/js/client.ts#L32).
In addition, after setting the transport layer for the sentry (native) client, it overrides external [`options.transport`](https://github.com/getsentry/sentry-react-native/blob/main/src/js/client.ts#L35) definition, so the `options` sent to `super(options)` (The `BaseClient` as defined in [`@sentry/core`](https://github.com/getsentry/sentry-javascript/blob/master/packages/core/src/baseclient.ts#L108)) are actually not as given in the `Sentry.init(...)`

The real issue is that `@sentry/react-native` is not checking for external `transport` function coming from [`options: ReactNativeOptions`](https://github.com/getsentry/sentry-react-native/blob/c905df1f9cecfc490d72a23182286408c6d57aa3/src/js/options.ts#L108) which I think is a bug as they need to support or according to their docs.

I will open a bug/PR for Sentry guys and let's see what their opinion on this

**Comment by zivl** [comment]
a PR has been created: https://github.com/getsentry/sentry-react-native/pull/2398

**Comment by liamjones** [comment]
Thank zivl. Separately, I've confirmed the types are incorrect for V7 and have opened PR #116 to resolve that.

**Comment by zivl** [comment]
a fix has been released in [`@sentry/react-native@4.2.2`](https://github.com/getsentry/sentry-react-native/blob/main/CHANGELOG.md#422)
